### DISCUSSION TITLE
Original discussion recommending a Spanish version of the Engram layout

### DISCUSSION THREAD

**Discussion by binarybottle** [binarybottle/engram-es#3] [Ideas]
Copied from https://github.com/binarybottle/engram/issues/25 

NickG13 commented 19 days ago • 
Spanish is one of the most spoken languages in the world. Yet, we really lack a specialized kw layout. We dont have any. Even the DVORAK variant are clueless of our letter frequency.
Would you be capable of doing a spanish engram variant, following all the same method?
You would make millions of typers very happy
Thanks

@binarybottle
 
Owner
binarybottle commented 19 days ago • 
I sure could! Could you identify the following resources?:

Large, accurate, and representative Spanish letter and bigram frequency counts, like Peter Norvig's for English: http://norvig.com/mayzner.html
Spanish punctuation frequency counts, like this example for English: http://www.viviancook.uk/Punctuation/PunctFigs.htm
#1 could potentially be generated from a large and representative Spanish text corpus comparable to what I assembled for English text: https://github.com/binarybottle/text_data, either with my python code, or with available tools like https://onlinecryptotools.com/analyze-letter-frequency

@binarybottle
 
Owner
binarybottle commented 19 days ago
How does one deal with diacritical marks for the seven additional letters?: Á, É, Í, Ñ, Ó, Ü and
(https://www.sttmedia.com/characterfrequency-spanish). Should these be accessed using key binding with some other key besides the Shift key, such as the Alt-Right key?

@NickG13
 
Author
NickG13 commented 19 days ago • 
You have an awesome attitude, thank you for helping out spanish typers!
I was trying to create one on my own, but I think your system is a lot better and all I want is to help and hopefully reach a end point where we have a great spanish layout.

I will reply to you in order:
That said, here is some data you need, for several sources:

Letter frequency (I've seen it's actual legit data, I been searching for literally days for the best source and this is the most complete): https://www.sttmedia.com/characterfrequency-spanish I would add up all the accents % to calculate the real frequency of the accent ( ` )
Bigrams, trigrams... here is a range of different options
https://www.ngrams.info/spanish.asp
Some reflections and top bigrams and other interesting stuff
Maybe we could use some actual online-spanish texts in http://patorjk.com/keyboard-layout-analyzer if the other samples are not enough

On traditional keyboard layout, we accent characters with a certain key, as you may know. For example, to write "é" we first put " ´ " and then "e". I've been talking with a colleague and he seems to think like you: diacritical marks could get a more rapid access with other keys, but I don't consider it to bee urgent, personally. We already use a lot of key combinations to use apps and games, so maybe that could became counterproductive while writing in those apps with that new layout. You could try making two versions (one with usual accent system and another with key bindings) but maybe it's complicating things too much.
@NickG13
 
Author
NickG13 commented 19 days ago • 
This is some actual ranking I've found in my research (sorry, can't recall where I found it right now)

E
A
O
S
R
N
I
D
L
C
T
U
M
P
B
.
,
G
V
Y
Q
H
F
Z
J
Ñ
X
K
W

So, yes, comma and point are very used

@NickG13
 
Author
NickG13 commented 19 days ago
I've been looking at http://www.viviancook.uk/Punctuation/PunctFigs.htm and want to give some extra info. As per punctuation points, we could theorically consider the following:
. full stop = Phrases and words tend to be larger in spanish than english, so full stops are not as frequent.
, comma = The same happens with comma
; semi-colon = semi-colon are almost never used in spanish nowadays, only maybe for lists.
: colon = colons are more used than semi-colons, but also less used than in English IMHO
! exclamation = important! Exclamations marks always end the phrase, but these always should begin with "¡" to be correct gramatically (not everyone uses them, but they are important)
? question = the same happens with question pharses, that will begin with ¿
as for ! or ?, I think they have almost the same frequency than in english
’ apostrophe/ single quotation = there are some variations of spanish that use apostrophe a lot, like Catalan, but spanish doesn't
“ double quotation = the same use as on english, maybe less

hyphen = we don't use it as much as in english
@NickG13
 
Author
NickG13 commented 19 days ago
As per text lenght - words lenght, I don't have any study but there is some reflections on experienced people (and I can recall as an experienced writer)
https://forum.wordreference.com/threads/word-length-in-english-vs-spanish.254504/
https://www.reddit.com/r/Spanish/comments/3hb8zx/does_spanish_language_on_average_have_longer/

Spanish seems to occupy almost 25% more space

@Lobo-Feroz
 
Lobo-Feroz commented 19 days ago • 
Hello there!

I'm another spanish speaker interested in an optimised layout for spanish. Been researching a lot these last few years but didn't find any convincing alternative.

@NickG13 referred me to this thread. Engram was one of the algorithmically optimised layouts that I had investigated. I specially like that you kept no restrictions. None of that nonsense of keeping zxcv for windows shortcuts, or not moving around the punctuation.

I also want to thank you for extending your project to help spanish writers. There are really not many alternatives for a spanish keyboard, only adaptations from old human-designed layouts like dvorak. This might be a very helpful addition.

I will try to help where I can.

@Lobo-Feroz
 
Lobo-Feroz commented 19 days ago • 
@NickG13 in regards to this:

! exclamation = important! Exclamations marks always end the phrase, but these always should begin with "¡" to be correct gramatically (not everyone uses them, but they are important)
? question = the same happens with question pharses, that will begin with ¿

I think that the opening exclamation mark '¡' shoud be the shift character to '!'. (and also, '¿' should be the shift for '?')
Why? Because the major use for shift is to change to majuscules. Where do you do that most of the time? At the beginning of a sentence. Just where you will use '¡' and '¿'

There is an example already on the spanish keyboard. The <> key. However it's upside down. Unshifted it's < (beginning) and shifted is > (end). Yes, the qwerty spanish layout is weird.

@Lobo-Feroz
 
Lobo-Feroz commented 19 days ago • 
Ok, my take on this one:

How does one deal with diacritical marks for the seven additional letters?: Á, É, Í, Ñ, Ó, Ü and
(https://www.sttmedia.com/characterfrequency-spanish). Should these be accessed using key binding with some other key besides the Shift key, such as the Alt-Right key?

The name in spanish for all of these diacritical marks is "tilde" (https://en.wikipedia.org/wiki/Tilde). Currently the writing for all of those is differentiated.

For simple tildes ÁÉÍÓÚ, as @NickG13 said, it's the key '´' followed by the vowel key.

The two dots is called "Diaeresis" (https://en.wikipedia.org/wiki/Diaeresis_(diacritic)) and we have a key '¨' for that. Actually, it's the shift for the simple tilde key '´' (which, imho, it's not that bad). You can add these two dots to any vowel ÄËÏÖÜ äëïöü but it's only correct in the u. The rest are not used in proper spanish (but we can type them with our current keyboards).

The tilde for the Ñ is called "virgulilla" (weirdest name) and it's consireded a tilde as well. In qwerty-es we have a full key in the home row for the ñ, which I think it's very inefficient, as ñ has a frequency of 0.17% according to one of @NickG13 's links above. I think it could just be a modification on the n key: '´' + n = ñ

Now, here, I don't know what would be the best way to implement this tilde key. On qwerty-es, it's sequential, press tilde then press the vowel. An alternative could be simultaneous press with the opposite hand, but AltGr is already used for the vowel "e": AltGr + e = € (but just this one, 'aioun' don't have AltGr uses). Also, having a single key to switch to "tilde layer" would mean we'd need another method for the diaeresis '¨'.

And that's about all I have thought about tildes so far.

@NickG13
 
Author
NickG13 commented 17 days ago • 
Just to update you guys,
I've been doing some test with Engram in Keyboard layout analyzer (the updated version, not the one that is uploaded there) and Engram has a HUGE advantage over all the other popular layouts, even Dvorak spanish variant (Jorge Sanz) and such.
The text used is mine, from three articles on https://nicolasgutierrez.com/ , which I've seen (afterwards) it reflects perfectly the spanish letter frequency described on the links we shared.
I'm excited about this being done and just want to give some extra info and support.
Here are some screenshots
image

image

image

image

Of course , these are some random numbers from an automated website. I think it doesn't count accents (tildes) (´) or diaéresis (¨) but it's a fact that Engram seems the BEST option nowadays for spanish typing. Maybe a little optimization will make it UNBEATABLE.

image

@NickG13
 
Author
NickG13 commented 17 days ago • 
Just for the laughts, I've been trying to follow your system but it seems there's something I'm missing (or maybe the analysis system is not correct). I've found a way of having some layouts that feel more ergonomic but they seem to get a lot less points than your Engram layout. It's pretty logical since it seems your layout is faster but the program http://patorjk.com/keyboard-layout-analyzer/ doesnt seem to calculate ergonomics. Here they are:

v1
image
heatmap
image

v2
image
heatmap
image

v3
image
heatmap
image

Engram v2
heatmap
image

image

@NickG13
 
Author
NickG13 commented 16 days ago
Just saw I can share the URL
I've tested a little bit more, so the layouts are a bit different in some cases
http://patorjk.com/keyboard-layout-analyzer/#/load/kVFJVVLs

@NickG13
 
Author
NickG13 commented 16 days ago • 
Comparing layouts a little more to take more ideas, I've revised Alice in the wonderland heatmap for Engram
image

This remembered me to your first criteria, which makes most common letters in ^ form, not / like I was using.
image
Then, I've rearranged a few letters and got this heatmap with spanish text:
image
I also have taken account of your other ergonomic criteria, the most difficult were:
7. Avoid stretching shorter fingers up and longer fingers down.
8. Avoid using the same finger.
Also taken into account most common bigrams, trying to avoid point 7 and 8, and making the movement being inwards, not outwards.
DE 2088635
ES 1882962
EN 1850294
EL 1638128
LA 1464005
OS 1460757
ON 1307790
AS 1268034
ER 1239281
RA 1197125
AD 1167809
I really like this layout, let's see what you think

@NickG13
 
Author
NickG13 commented 13 days ago • 
@binarybottle Spanish corpus found https://wortschatz.uni-leipzig.de/en/download/Spanish huge database
Do you have any software that gets bigrams and monograms frequency?
We want to help you progress faster, if that's possible, count with us

@binarybottle
 
Owner
binarybottle commented 8 days ago
@NickG13 -- Thank you very much for sharing your experiments, the keyboard layout analyzer results, and the link to the Spanish corpus! I have downloaded all of the largest news corpora from each year in the top section, and the latest Wikipedia corpus.

@Lobo-Feroz -- Thank you for following up on the diacritical marks! I think I have an idea for how to accommodate these in and intuitive manner if I use an extra key like this Shift key.

I really appreciate your enthusiastic support. I am currently preoccupied with an impending deadline for a conference, but hope to begin tackling this next weekend…

@NickG13
 
Author
NickG13 commented 8 days ago
@NickG13 -- Thank you very much for sharing your experiments, the keyboard layout analyzer results, and the link to the Spanish corpus! I have downloaded all of the largest news corpora from each year in the top section, and the latest Wikipedia corpus.

@Lobo-Feroz -- Thank you for following up on the diacritical marks! I think I have an idea for how to accommodate these in and intuitive manner if I use an extra key like this Shift key.

I really appreciate your enthusiastic support. I am currently preoccupied with an impending deadline for a conference, but hope to begin tackling this next weekend…

Sure, deadlines are first! Thank you for the update!

@Lobo-Feroz
 
Lobo-Feroz commented 8 days ago
@binarybottle - thanks for your answer!

once you beging working on this layout (no hurries here), please don't hesitate to send us some grunt work. I'll be happy to help where I can.

@binarybottle
 
Owner
binarybottle commented 21 hours ago
@NickG13, @Lobo-Feroz, @sunaku -- I have created a new repo, setting up a Spanish version of the Engram layout, and am now crunching initial layouts:

https://github.com/binarybottle/engram-es

Let's continue our discussion there!

@binarybottle binarybottle closed this 21 hours ago

**Comment by iandoug** [comment]
http://patorjk.com/keyboard-layout-analyzer/#/load/31nf5cXb

Added some diacritics. 
Need to see frequency charts for letters and bigrams ...

**Comment by binarybottle** [comment]
I'm using the frequencies from here: http://practicalcryptography.com/cryptanalysis/letter-frequencies-various-languages/spanish-letter-frequencies/

**Reply by iandoug** [reply]
I do upper and lower separately... plus non-letters :-)
Will see how much cleaning those Leipzig corpus files need tomorrow.

**Reply by Lobo-Feroz** [reply]
@iandoug I'm assuming these are the corpora you use:

https://wortschatz.uni-leipzig.de/en/download/Spanish

I remember getting this link from you on reddit. I'm gonna check some of these to see if I can find any other rare symbols.

**Reply by iandoug** [reply]
What I posted on Reddit was pulled from one of those files. I grabbed the others last night (biggest version of each, but ignored the South American ones) and will unpack and take a look later today.
South America may be using ANSI rather than ISO, or have special words/chars that will complicate things further.

**Reply by iandoug** [reply]
I have scripts that I used to pull books from Gutenberg for the English corpus, will see if they can be persuaded to do Spanish.
What I did for English was to first clean the texts (strip Gutenberg text as well) then if file was over x kb, take an extract after skipping first 200 lines (to get past "Contents" etc). This produced chunks of 10kB or so from many different authors and styles, and avoids the Tarzan/Zorro/James Bond problem ... where normally rare letters are used excessively.

Leipzig files are 'sentenced', so I need to artificially add paragraphs.

Suggestion:
After 100 words or 3 sentences, there is a 50% chance of adding paragraph.
After 4 sentences, there is a 60% chance of adding paragraph.
After 5 sentences, there is a 70% chance of adding paragraph.
After 6 sentences, there is a 80% chance of adding paragraph.
After 7 sentences, there is a 90% chance of adding paragraph.
After 200 words or 8 sentences, there is a 100% chance of adding paragraph.

Comments welcome. :-)

**Reply by iandoug** [reply]
> @iandoug I'm assuming these are the corpora you use:
> https://wortschatz.uni-leipzig.de/en/download/Spanish

This is best free source I have found for diverse languages.

**Reply by binarybottle** [reply]
@iandoug -- Is there a problem with the bigram frequencies I'm using to generate layouts?  Do you think it's non-representative?

**Reply by iandoug** [reply]
@binarybottle  ... I don't know. Busy trying to get single char frequencies at the moment, will upload when done. Will then look at bigrams ... need to make sure Unicode is handled properly. Existing programs were for US ANSI, which is easier.

Freq lists for encryption have different goals to keyboard layouts.

**Reply by iandoug** [reply]
Okay, your source does not include the space character. So their percentages are higher.

**Reply by binarybottle** [reply]
I don't really care about the space character, as it can be typed with either hand.

**Reply by iandoug** [reply]
Which hand you use affects the score on KLA. Hand use metric, and hand alternation.

**Reply by binarybottle** [reply]
@iandoug, I care about what's most comfortable to type, not how KLA scores!

**Comment by iandoug** [comment]
Extracts from corpus:
6 hbitos saludables para emprendedores (etc ...)
6. La pgina web hace que las ventas aumenten El aumento (etc...)
7. 6 lmites profesionales que el Jefe  (etc...)

I assume hbitos, pgina and lmites are typos ... hope there's not too many nasties like that, will mess up the analysis.

**Reply by Lobo-Feroz** [reply]
Hey @iandoug, this is similar to the chinese character that showed up in "sorprendió" instead of ó.

All these words are missing tilded vowels:

- "hbitos" should be "hábitos" _(habits)_
- "pgina" should be "página" _(page)_
- "lmites" should be "límites" _(limits)_

These are not typos, these are failures of the corpus to detect or process tilded vowels. It's consistent.

**Reply by iandoug** [reply]
Mmmm... so if this is a problem with their corpus, assuming related to either Unicode issues or source being something like &aacute; which got discarded....
That's annoying.
Will have to see if I can get anything from Gutenberg, then process separately, and compare frequencies...

**Reply by iandoug** [reply]
I thought this may be limited to certain files but seems like pervasive problem. For example, the files contain both página and pgina. 
I can fix that, the problem is how to find all/most cases of missing letters .... 

Okay, got Linux Spanish word list, let me see if we can find bad words in corpus.

Will revert.

**Reply by iandoug** [reply]
This is just depressing. That corpus is a mess.

Eg.
 deportivassituacinalojamientospadietasinstalacionesespecial

Are these words?
ciclistascomplejos deportivosnuestros clientessolicitar

There are tons of unknown words, will have to do this slowly ....
![2021-08-02_11-24](https://user-images.githubusercontent.com/12673886/127838067-c624916f-f18f-4ed9-98e6-88ca57f15f19.png)

**Reply by Lobo-Feroz** [reply]
> deportivassituacinalojamientospadietasinstalacionesespecial
> Are these words?

deportivas (the femenine form of the adjetive "from sports" - "sporty")
situación (situation, missing the ó)
alojamientos (accommodation)
padietas ¿¿¿ ??? ("dietas" means "diets", "padietas" is not a word)
instalaciones (installations)
solicitar (to ask for)

> ciclistascomplejos deportivosnuestros clientessolicitar

ciclistas (cyclists)
complejos (complexes)
deportivos (from sports, sporty)
nuestros (ours)
clientes (clients)
solicitar (to ask for)

> That corpus is a mess.

Yeah, agreed!

**Reply by Lobo-Feroz** [reply]
About the image, haven't read all the words but it seems it has the same problem, tilded letters are missing.

For instance there is "espaa" with is obviously the name of our country Spain (España) without the ñ.

Weird indeed. One would think that a corpus collection for a language should gather correctly the special chars for that language

**Reply by iandoug** [reply]
Checking against Linux Spanish word list. Still tweaking it, like dropping last s to see if that exists.
Also started list of 'loanwords' like yahoo, twitter, etc.

Am first getting a feel for the scope of the problem ... it may be the same x00 words repeated over and over.

Those words above, eg ciclistascomplejos, should that be two words or is Spanish like Afrikaans or Dutch where you can string descriptive words and nouns together?

Cheers, Ian

**Reply by Lobo-Feroz** [reply]
No, in spanish you can't string along different words. Unless the word itself is already existing as a combination of two other words in the dictionary. For instance umbrella is "paraguas", literally "water-stop". But you can't invent a new one on the fly, so to speak.

So "ciclistascomplejos" should be "ciclistas complejos" ("complex cyclists"). Which, btw, doesn't make a lot of sense, how is a cyclist something complex? :P

**Reply by iandoug** [reply]
Thought it might mean complex bicycle, or 'some sort of building where people do cycling-related activities' ... 'complex' has different meanings in English, assumed it might be a case of language pollution / mis-translation into Spanish. We have that problem a lot here where English words are forced into other languages.

**Reply by Lobo-Feroz** [reply]
Maybe within the context, considering the sentence or paragraph, we might see if some of these examples are typos, inaccuracies or just garbage. Ofc, only if you think it's relevant to know.

**Reply by iandoug** [reply]
> Checking against Linux Spanish word list. 

which only has 86k entries ... finding more words NOT on it ...

Plan B: getting word lists from www.listapalabras.com ... will take a while to get whole alphabet (currently at C) but has vastly more words. Will merge these with what is on Linux list.

**Reply by iandoug** [reply]
Ok, so my 'unknown words' list is 1.9 million long. Entries from about 800k to the end are single occurrences, e.g.
siluetada	2
videnteadivinamísticaproféticapsíquica	2
incompletas;	2
raketa	2
chalandri	2
desmalezados;	2
rougeen	2
trasplantársela	2
pkill	2
fedezek	2
tontaína	2
vppb	2
revori	2
fortalenciendo	2
lÉcotai	2
jaaaaaa	2
bridalplasty	2
khural	2
yúne	2
bernuéz	1
tridiavo	1
pepelito	1
entradascumple	1
maquibar	1
lecocarpu	1
desonco	1
vim+ant+script	1

The top few are (all lower-cased, I should check as-is plus lower). Have taken out the place names.
millone	407845
sólo	330610
cómo	186122
aún	177302
paíse	177279
mujere	174375
última	160219
quiene	158136
dólare	140909
autoridade	133910
sociale	118471
jóvene	116276
trabajadore	109288
actividade	108780
the	108604
económica	105235
eleccione	98685
principale	91198
cuale	90164
mayore	85357
jugadore	84549
pequeña	80803
locale	78048
única	76351
pp	65958
próxima	64565
finale	64092

Are these all spelled correctly?
Don't know why so many "the"... seems to be English titles:
 El ttulo del film fue "Raiders of the Lost Ark - The Adaptation", que lleg a manos del mismsimo Steven Spielberg, quien les felicit por su trabajo.

I think if we can sort out the frequent words then I can just ignore any sentences with the others ... there are 26,329,994 lines so dropping 1 - 1.5 million is not a big deal.

Thanks, Ian

**Reply by binarybottle** [reply]
I strongly recommend tossing out any single entry words.

**Reply by iandoug** [reply]
I was intending to drop the entire sentence. Which may be overkill.
Currently reprocessing to see how many of the 26 million sentences have possibly unknown words. There's a lot of names (people, places) which are not technically "Spanish words" but would still get typed on occasion ...

Realised dropping all such lines would be a lot more than the number of distinct words.
The original aim of this exercise (which has taken all day) was to find "typos" that would mess up the bigram stats. But  sorting it out is a much bigger exercise than I expected..

FWIW, my "known" words list is  728,374 words... lots of conjugations.

**Reply by binarybottle** [reply]
Thank you for your diligence!  You may be reaching the point of diminishing returns.  If you have 26 million sentences with many millions of words, I would think that removing single-entry strings would not compromise the representativeness of the remainder.

**Reply by iandoug** [reply]
Found 18,810,258 bad lines out of 26,329,994 lines
"bad line" == at least one unknown word.

Busy rerunning to get lists+counts of good words as well as the bad, to make sure my lookups are not doing something stupid.
The 'bad words' are probably Spanish for the most part, but manually going through a list of 1.8 million is not a good idea. OTOH there can't be THAT many words so something is wrong.

We examine results and make a plan by tomorrow.

**Reply by iandoug** [reply]
Conclusion> My 728k word list sucks. Or I'm missing some aspect of Spanish word formation.

Found 1,940,663 distinct bad words and 175,015 distinct good words.

Let me sleep on it.

**Reply by iandoug** [reply]
UniLeipzig don't scrub very well. There is still HTML markup, and plain junk like this.

![2021-08-03_00-01](https://user-images.githubusercontent.com/12673886/127929565-d46cb600-407e-4048-aa7b-be9aa45a6f12.png)

**Reply by Lobo-Feroz** [reply]
> Conclusion> My 728k word list sucks. Or I'm missing some aspect of Spanish word formation.

@iandoug let me try to help. Let's go through that word list with big frequencies.

These are **valid spanish words**:

sólo 330610 (only) 
    the canoninal form at this time is "solo" (no tilde), but "sólo" is the old version, and it's accepted
cómo 186122 (how)
aún 177302 (yet)
última 160219 (last, feminine gender)
económica 105235 (economic, feminine gender)
pequeña 80803 (small, feminine)
única 76351 (small, feminine)
próxima 64565 (nearby, feminine)

These are **plural words, missing the final "s"**:
For instance, "million" in spanish is "millón", "millions" would be "millone**s**". "millone" is not a correct word, but it's clearly "millones" without the final s. All these words have this missing last 's':

millone 407845 --> should be "millones" (millions)
paíse 177279 --> "paises" (countries)
mujere 174375 --> "mujeres" (women)
quiene 158136 --> "quienes" (who, plural)
dólare 140909 --> "dólares" (dollars)
autoridade 133910 --> "autoridades" (authorities)
sociale 118471 --> "sociales" (socials)
jóvene 116276 --> "jóvenes" (young people)
trabajadore 109288 --> "trabajadores" (workers)
actividade 108780 --> "actividades" (activities)
eleccione 98685 --> "elecciones" (elections)
principale 91198 --> "principales" (main ones)
cuale 90164 --> "cuales" (which, plural)
mayore 85357 --> "mayores" (bigger / older, plural)
jugadore 84549 --> "jugadores" (players)
locale 78048 --> "locales" (locals)
finale 64092 --> "finales" (ends)

**Special cases**: 

the 108604 - as you say, this is most probably from an english name, such as of a movie
pp 65958 - in smallcaps it's not a word. in caps PP is the common way to refer to the Popular Party, one of the two biggest political parties in Spain

Hope this helps,

**Reply by iandoug** [reply]
> aún 177302 (yet)
> última 160219 (last, feminine gender)

Thanks. Am adding your corrections.
I was dropping the final s to see if I have that version. Will also check without final es. Is that general way of making a word plural?

Re "aún", I find 
aúna	438
aúnan	144
aúnas	1
aúne	44
aúnen	28
aúnes	1
aúno	7

Are these gender/plural versions?
Is Spanish like French where every inanimate object has a gender or is the gender of the object linked to the person in the sentence, or something else?
I'm basically looking for other "transformations" I can do to words to check if their 'root' form exists. Like add or remove a or o.

Thanks.

**Reply by Lobo-Feroz** [reply]
> I was dropping the final s to see if I have that version. Will also check without final es. Is that general way of making a word plural?

Yes. If the word ends in "a" or "o", it adds an "s". If it ends in a consonant, it adds "-es".

Actually it's a little more complex than that, so if you want to make a generic algorithm I can write you the whole plural rules. But maybe it's faster to check a list.



> Re "aún", I find
> aúna 438
> aúnan 144
> aúnas 1
> ...
 
> Are these gender/plural versions?

Worse, this is a _verb_. 

Verb conjugation in spanish is a bit tricky compared to other languages.

Here's the conjugation for this verb, "aunar" (to gather): https://www.wordreference.com/conj/EsVerbs.aspx?v=aunar

There are a lot of irregular verbs in spanish so I think it would be very complex to algorithmically check them all. IMHO this is a shortcoming of the dictionary.

> Is Spanish like French where every inanimate object has a gender or is the gender of the object linked to the person in the sentence, or something else?

Each object has a gender, like in german, but with only 2 genders instead of 3.

> I'm basically looking for other "transformations" I can do to words to check if their 'root' form exists. Like add or remove a or o.

Yeah sometimes that might work, but not always. I wonder why the dictionary doesn't have these variations.

Also, I propose we move this to the "cleaning the corpus" thread.

**Reply by iandoug** [reply]
> All these words have this missing last 's':

My fault, I dropped the s to test for singular, and then printed the chopped version instead of original.
Was originally
millones	408694
sólo	330572
cómo	186117
países	177491
aún	177345
mujeres	174811
quienes	158108
dólares	141507
autoridades	133954
sociales	118732
jóvenes	116469
trabajadores	109465
actividades	108814
the	108660
última	107151
elecciones	98792
principales	91233
cuales	90138
mayores	85480
jugadores	84630
económica	78841
locales	78218
única	69479
pp	65862
finales	63960
lugares	62861
menores	62576
familiares	60599
quién	58735
eeuu	58566
miles	58380
ciudades	58303
instituciones	58122
sectores	57596

**Reply by Lobo-Feroz** [reply]
All these words look legit to me.

PP should be capitalized, but I suppose some people just write "pp" for ease.
